### PR TITLE
Mimic vanilla CS behaviour with NPC stats auto calculations.

### PR DESCRIPTION
**Pull request by cc9cii**
Should resolve Bug #2663.

~~- Currently does not update npc autocalc values when race, class or skills tables change.~~

### CODE REVIEW

**Comment by scrawl** [comment]
Still missing the spell autocalc. 

I'm not too happy about the duplicated code. Getting the autocalc to function correctly in the first place was a nightmare. Problems like rounding errors or the iteration order of records can completely break the results. I'd like to have the code in a central location, so it's easier to ensure that it doesn't break. I have a unit test comparing the results with vanilla results provided by Hrnchamd. 

Only problem with using the existing code is exchanging the OpenMW store with the OpenCS store. We could define a common interface for the Spell and GMST stores that OpenMW and OpenCS then hook into. 

Just make sure the OpenCS Spell store iterates through records in the order that the content files define them, else the autocalc isn't going to work correctly.

**Comment by cc9cii** [comment]
Ok, will have a look.

**Comment by cc9cii** [comment]
> Still missing the spell autocalc.
> ....
> Just make sure the OpenCS Spell store iterates through records in the order that the content files define them, else the autocalc isn't going to work correctly.

AFAIK, OpenCS does not load files according to the launcher order - which means the autocalculated spells may be different b/w OpenMW and OpenCS.

Am i missing something here?

**Comment by zinnschlag** [comment]
OpenCS does load files according to the load order.

**Comment by cc9cii** [comment]
Autocalc spells done:
autocalc spells

Some callouts:
* The component name GamePlay isn't so great, but can't think of a better one (formulae maybe?)
* Had to move SharedIterator as part of the shared interface with OpenCS - not sure if there is another way?

File: components/CMakeLists.txt [hunk 1]
```diff
@@ -161,14 +165,14 @@ include_directories(${BULLET_INCLUDE_DIRS} ${CMAKE_CURRENT_BINARY_DIR})
...
+target_link_libraries(components
     ${Boost_SYSTEM_LIBRARY}
     ${Boost_FILESYSTEM_LIBRARY}
     ${Boost_THREAD_LIBRARY}
     ${Boost_PROGRAM_OPTIONS_LIBRARY}
     ${Boost_WAVE_LIBRARY}
     ${OGRE_LIBRARIES}
-    ${OENGINE_LIBRARY}    
+    ${OENGINE_LIBRARY}
     ${BULLET_LIBRARIES}
```
Comment (scrawl) [inline components/CMakeLists.txt]: Please undo the whitespace changes in this file. I smell merge conflicts.

File: apps/openmw/mwclass/npc.cpp [hunk 2]
```diff
@@ -58,116 +62,47 @@ namespace
...
-        bool male = (npc->mFlags & ESM::NPC::Female) == 0;
+    public:
 
-        int level = creatureStats.getLevel();
-        for (int i=0; i<ESM::Attribute::Length; ++i)
-        {
-            const ESM::Race::MaleFemale& attribute = race->mData.mAttributeValues[i];
-            creatureStats.setAttribute(i, male ? attribute.mMale : attribute.mFemale);
-        }
+        Stats(MWMechanics::CreatureStats& creatureStats, MWMechanics::NpcStats& npcStats)
```
Comment (scrawl) [inline apps/openmw/mwclass/npc.cpp]: You don't need both CreatureStats and NpcStats. NpcStats is derived from CreatureStats (although it wasn't at the time the original autocalc was written).

**Comment by scrawl** [comment]
For a component name, ```autocalc``` perhaps?

Moving StoreBase and SharedIterator is overkill IMO. Most of its interface is irrelevant. All we need is:

- find a GMST
- find a skill
- find a MagicEffect
- list all spells

That is, the shared interface should be at most 4 functions.

File: apps/openmw/mwclass/npc.cpp (hunk 2 above)
Comment (cc9cii) [inline apps/openmw/mwclass/npc.cpp]: I couldn't get that to work for both OpenMW and OpenCS, unfortunately.

**Comment by cc9cii** [comment]
As stated above, I couldn't find a way to get a common interface without making other invasive changes (which I didn't know if it was ok to do - for example the loop where ESM::Spell* is iterated)

File: apps/openmw/mwclass/npc.cpp (hunk 2 above)
Comment (scrawl) [inline apps/openmw/mwclass/npc.cpp]: Where is the problem? What I'm suggesting is simple. You remove mCreatureStats, then change code accessing mCreatureStats to use mNpcStats instead. I don't see why that would affect OpenCS.

**Comment by scrawl** [comment]
The function for getting spells could look like this:

```
void listSpells(std::vector<ESM::Spell*> spells);
```

then in the autocalc code, you do:

```
std::vector<ESM::Spell*> spells;
storeInterface->listSpells(spells);

for (std::vector<ESM::Spell*>::const_iterator it = spells.begin(); ....
```

**Comment by cc9cii** [comment]
Ok, will try that.

File: apps/openmw/mwclass/npc.cpp (hunk 2 above)
Comment (cc9cii) [inline apps/openmw/mwclass/npc.cpp]: I don't follow.  Do you mean to change the parameter here to use NpcStats& instead?
```
void autoCalculateAttributes (const ESM::NPC* npc, MWMechanics::CreatureStats& creatureStats)
```

File: components/CMakeLists.txt (hunk 1 above)
Comment (cc9cii) [inline components/CMakeLists.txt]: Will do.  Perhaps you can cleanup the tabs after your merge.

File: apps/openmw/mwclass/npc.cpp (hunk 2 above)
Comment (scrawl) [inline apps/openmw/mwclass/npc.cpp]: > Do you mean to change the parameter here to use NpcStats& instead?

Exactly.

**Comment by cc9cii** [comment]
How do you get std::vector<ESM::Spell*> from MWBase::Environment::get().getWorld()->getStore().get<ESM::Spell>() without iterating and making copies?

File: apps/openmw/mwclass/npc.cpp (hunk 2 above)
Comment (cc9cii) [inline apps/openmw/mwclass/npc.cpp]: Not that I look at it, that was pretty obvious.  Not sure what I was thinking.

**Comment by scrawl** [comment]
```
for (MWWorld::Store<ESM::Spell>::iterator iter = spells.begin(); iter != spells.end(); ++iter)
     spellVector.push_back(&*iter);
```

**Comment by cc9cii** [comment]
That's still making copies each time (seems inefficient), but if you're ok with it I'll implement it that way.

**Comment by cc9cii** [comment]
How about adding one method to MWWorld::Store instead?  It exposes mShared, but avoids looping & copying.
```
         const std::vector<T *>& getShared() const {
             return mShared;
         }
```

**Comment by scrawl** [comment]
mShared is an implementation detail. We shouldn't rely on it being a vector.

> That's still making copies each time (seems inefficient), but if you're ok with it I'll implement it that way.

Copying the pointers, not the data. I wouldn't worry about it.

**Comment by cc9cii** [comment]
Oh for Pete's sake!  Too many distractions :-(

File: components/autocalc/store.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#ifndef COMPONENTS_AUTOCALC_STORE_H
+#define COMPONENTS_AUTOCALC_STORE_H
+
+#include <vector>
+#include <string>
+
+namespace Loading
+{
+    class Listener;
```
Comment (scrawl) [inline components/autocalc/store.hpp]: This forward declaration can be removed now.

File: components/autocalc/store.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+namespace ESM
+{
+    class ESMWriter;
+    class ESMReader;
```
Comment (scrawl) [inline components/autocalc/store.hpp]: And these two...

File: components/autocalc/autocalcspell.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+#include <components/esm/attr.hpp>
+#include <components/esm/loadmgef.hpp>
+#include <components/esm/loadrace.hpp>
+#include <components/esm/loadclas.hpp>
+#include <components/esm/loadspel.hpp>
+#include <components/esm/loadnpc.hpp>
+
+#include "autocalc.hpp"
+
+// Most of the code in this file was moved from apps/openmw/mwmechanics/autocalcspell.cpp
```
Comment (scrawl) [inline components/autocalc/autocalcspell.cpp]: Can we remove this comment (and the similar one in autocalc.cpp)? Anyone is free to browse the repository for history lessons.

File: components/autocalc/autocalc.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        for (int i = 0; i < 2; ++i)
+        {
+            int bonus = (i==0) ? 10 : 25;
+
+            for (int i2 = 0; i2 < 5; ++i2)
+            {
+                int index = class_->mData.mSkills[i2][i];
+                if (index >= 0 && index < ESM::Skill::Length)
+                {
+                    stats.setBaseSkill (index, stats.getBaseSkill(index) + bonus);
```
Comment (scrawl) [inline components/autocalc/autocalc.cpp]: This relies on all skills being initialized to ```0``` before the autocalc. I know that's the case in OpenMW, but not sure about OpenCS.

Better would be to have the autocalc overwrite the skills, rather than adding. Then we can also get rid of the getBaseSkill function.

For reference: https://wiki.openmw.org/index.php?title=Research:Stats_and_Levelling#Skills

File: components/autocalc/autocalc.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                    round_ieee_754(
+                            stats.getBaseSkill(skillIndex)
+                    + 5
+                    + raceBonus
+                    + specBonus
+                    +(int(level)-1) * (majorMultiplier + specMultiplier)), 100)); // Must gracefully handle level 0
+        }
+    }
+
+    unsigned short autoCalculateHealth(int level, const ESM::Class *class_, StatsBase& stats)
```
Comment (scrawl) [inline components/autocalc/autocalc.cpp]: Here StatsBase can be const.

File: components/autocalc/autocalc.hpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    public:
+
+        StatsBase();
+        virtual ~StatsBase();
+
+        virtual unsigned char getBaseAttribute(int index) const = 0;
+
+        virtual void setAttribute(int index, unsigned char value) = 0;
+
+        virtual void addSpells(std::string id) = 0;
```
Comment (scrawl) [inline components/autocalc/autocalc.hpp]: This is adding one spell, right? Then addSpell would be a better name.

Comment (scrawl) [inline components/autocalc/autocalc.hpp]: ```id``` can be passed by const reference.

File: apps/opencs/model/world/data.cpp [hunk 9]
```diff
@@ -1161,3 +1259,366 @@ void CSMWorld::Data::rowsChanged (const QModelIndex& parent, int start, int end)
...
+
+    const ESM::Race *race = &mRaces.getRecord(npc.mRace).get();
+    const ESM::Class *class_ = &mClasses.getRecord(npc.mClass).get();
+
+    bool autoCalc = npc.mNpdtType == ESM::NPC::NPC_WITH_AUTOCALCULATED_STATS;
+    short level = npc.mNpdt52.mLevel;
+    if (autoCalc)
+        level = npc.mNpdt12.mLevel;
+
+    CSMWorld::NpcStats *stats = new CSMWorld::NpcStats();
```
Comment (scrawl) [inline apps/opencs/model/world/data.cpp]: This object will leak if there are exceptions thrown by the autocalc.

**Comment by scrawl** [comment]
I am curious how the autocalc is triggered in OpenCS. Do we run it for every NPC in the game when the Objects table is opened? How long does that take? If it's too much of an issue, we could run it only when selecting "Edit" on an NPC.

**Comment by scrawl** [comment]
> If it's too much of an issue, we could run it only when selecting "Edit" on an NPC.

The autocalculated values aren't really model data, since they aren't saved with the file. So showing them in the table would be wrong IMO. @zinnschlag , any thoughts?

File: components/autocalc/store.hpp (hunk 3 above)
Comment (cc9cii) [inline components/autocalc/store.hpp]: Yes, leftover after removal.

File: components/autocalc/store.hpp (hunk 4 above)
Comment (cc9cii) [inline components/autocalc/store.hpp]: Ditto.

File: components/autocalc/autocalcspell.cpp (hunk 5 above)
Comment (cc9cii) [inline components/autocalc/autocalcspell.cpp]: Ok.

File: components/autocalc/autocalc.cpp (hunk 6 above)
Comment (cc9cii) [inline components/autocalc/autocalc.cpp]: I simply copied the code over.  In fact took some care in doing so, as I was under the impression that keeping things the same would be desirable (apparently false assumption I guess).

File: components/autocalc/autocalc.hpp (hunk 8 above)
Comment (cc9cii) [inline components/autocalc/autocalc.hpp]: Again, the code was simply copied over, keeping the names of the methods the same.

File: apps/opencs/model/world/data.cpp (hunk 9 above)
Comment (cc9cii) [inline apps/opencs/model/world/data.cpp]: Good point.  I have to think about how to get around this.

**Comment by cc9cii** [comment]
> I am curious how the autocalc is triggered in OpenCS. Do we run it for every NPC in the game when the Objects table is opened?

That is how I originally implemented it, but doing that too long.  There would be no real point, since we don't save the values, anyway.

> How long does that take? If it's too much of an issue, we could run it only when selecting "Edit" on an NPC.

That is basically what happens, except that changing GMST, Skill, Race powers, NPC level, etc, will re-trigger the calculation.

> So showing them in the table would be wrong IMO.

As the title of the PR says, this is mimicing how vanilla presents NPC data.

File: apps/opencs/model/world/data.cpp (hunk 9 above)
Comment (scrawl) [inline apps/opencs/model/world/data.cpp]: One idea:
<pre>
std::auto_ptr<CSMWorld::NpcStats> stats (new CSMWorld::NpcStats);

// ... perform autocalc ....

emit cacheNpcStats (npc.mId, stats.release());
</pre>

File: components/autocalc/autocalc.hpp (hunk 8 above)
Comment (scrawl) [inline components/autocalc/autocalc.hpp]: I can't find a function named addSpells in the original code.

File: apps/opencs/model/world/nestedcoladapterimp.cpp [hunk 10]
```diff
@@ -1069,23 +1070,66 @@ namespace CSMWorld
...
-                    cell.mAmbi.mFogDensity : QVariant(QVariant::UserType);
+            case 1:
+            {
+                if (isInterior && !behaveLikeExterior)
+                    return cell.mAmbi.mAmbient;
+                else
+                {
+                    UserInt i(cell.mAmbi.mAmbient);
+                    return QVariant(QVariant::fromValue(i));
+                }
```
Comment (scrawl) [inline apps/opencs/model/world/nestedcoladapterimp.cpp]: I'm afraid I don't understand what this does. 

Doesn't QVariant support int and float types by default?

Is there a difference between returning a QVariant(cellAmbient) and a QVariant(UserInt(cellAmbient))?

Previously, an empty QVariant would be returned if the field doesn't apply. Why was this changed?

File: components/autocalc/autocalc.hpp (hunk 8 above)
Comment (cc9cii) [inline components/autocalc/autocalc.hpp]: Actually, you are correct.  I morphed getSpells().add() to addSpells().  My bad.  Fixed.

File: apps/opencs/model/world/nestedcoladapterimp.cpp (hunk 10 above)
Comment (cc9cii) [inline apps/opencs/model/world/nestedcoladapterimp.cpp]: Previous implementation (returning QVariant() only) resulted in the values in the dialogue subview to be displayed as the default, usually zero.  There needs to be another way to send values but still disable editing, grey it out, etc.

File: apps/opencs/model/world/data.cpp (hunk 9 above)
Comment (cc9cii) [inline apps/opencs/model/world/data.cpp]: Thanks, I'll use it ;-)

File: components/autocalc/autocalc.cpp (hunk 6 above)
Comment (scrawl) [inline components/autocalc/autocalc.cpp]: > I was under the impression that keeping things the same would be desirable

The original skills aren't used, so the presence of ```getBaseSkill``` is just misleading. Your commit didn't introduce this problem, but it did increase its exposure by moving to a virtual function that now has to be implemented in multiple places. So let's go the extra mile for cleanup.

Don't worry too much about introducing regressions. I have a text file with the vanilla autocalc results for all NPCs in the game, and will be double checking with that before merge. If there are new bugs, we'll find them.

File: apps/opencs/model/world/nestedcoladapterimp.cpp (hunk 10 above)
Comment (scrawl) [inline apps/opencs/model/world/nestedcoladapterimp.cpp]: I see the problem. I'm not familiar enough with Qt though to suggest better solutions. @zinnschlag, any ideas?

**Comment by scrawl** [comment]
> That is how I originally implemented it, but doing that too long. There would be no real point, since we don't save the values, anyway.
> That is basically what happens, except that changing GMST, Skill, Race powers, NPC level, etc, will re-trigger the calculation.

Good to know. I think there is a bit of excess code left over from the previous implementation. I see about 300 lines of code just for keeping track of data changes. We don't need so fine grained checks, if the autocalc is only being run for one NPC at a time anyway. And the global cache shouldn't be necessary anymore.

**Comment by cc9cii** [comment]
I will see what I can clean up, but some of them will need to stay.  This is because in OpenCS the values can change dynamically - i.e. a Skills table may be open at the same time as the NPC dialogue.  Changing skill attribute should automatically update the dialogue values/spells (without traversing the entire table each time, hence the checks)

File: apps/opencs/model/world/nestedcoladapterimp.cpp (hunk 10 above)
Comment (cc9cii) [inline apps/opencs/model/world/nestedcoladapterimp.cpp]: I think this is less invasive than returning a struct with a bool and an int, for example.  Such a change will mean a significant amount of change elsewhere in OpenCS.

Comment (zinnschlag) [inline apps/opencs/model/world/nestedcoladapterimp.cpp]: At least the User types would have to be renamed. UserX really does not explain the purpose well. Also, this is a separate issue and shouldn't be handled within this pull request which is already messy enough.

But looking at the code we are seeing here, is there really a need to make changes? Do we really need a way to display a greyed out value? Being greyed out means here that it does not apply to the record in question. Then why show it at all?

**Comment by zinnschlag** [comment]
Well, this is messy. I will state my opinion again that we should drop this feature completely. As mentioned before with the de-hardcoding planned after 1.0 most of the changes here need to be thrown out anyway. And depending on how it goes there is a chance that the feature will become flat out impossible to implement in OpenMW-CS at some point.

Anyway, if we are going ahead with this, maybe a different approach should be attempted. The autocalculated data fields are fundamentally different from the editable one. It seems that at least a part of the mess comes from the attempt to merge these together.

Maybe have a separate section in the NPC dialogue subview that is used for auto-calc values? This would remove the need for things like UserInt and probably some other stuff.
If we worry that this will inflate the NPC dialogue subview too much, we could add a user setting that enables/disable showing of auto calculated values.

I know that is not how TES-CS does it. But I would like to remind everyone again that we are not re-implementing TES-CS. We are making a completely new editor. How things are done in TES-CS should be of no relevance to us. In fact there is no requirement to replicated TES-CS features to begin with.

**Comment by cc9cii** [comment]
Ok, understood.

**Comment by scrawl** [comment]
> Well, this is messy. I will state my opinion again that we should drop this feature completely. As mentioned before with the de-hardcoding planned after 1.0 most of the changes here need to be thrown out anyway. And depending on how it goes there is a chance that the feature will become flat out impossible to implement in OpenMW-CS at some point.

I don't agree with rejecting this feature. I think you are thinking too far ahead with the scripting concerns. Too many unknown variables to know for sure if that's going to be a problem. In the worst case, we can add a warning icon to the autocalc checkbox stating that the preview might not be accurate if a custom script is used.

That said, this PR does add a <i>lot</i> of code and I wasn't nearly done reviewing all of it. I am okay with shelving this feature for now on grounds of too much work for too little benefit. We probably have more urgent things to do.

**Comment by zinnschlag** [comment]
> I think you are thinking too far ahead with the scripting concerns. 

This will start to become relevant (at least to some degree) immediately after the release of 1.0. The de-hardcoding is one of the first post-1.0 features I plan to work on (after some general groundwork is done). Doesn't sound very far ahead to me.

> In the worst case, we can add a warning icon to the autocalc checkbox stating that the preview might not be accurate if a custom script is used.

That would be a very limited solution. Instead we can have one that works reliably. Fire up OpenMW with a debug profile that sets up the situation you want to review. Inspect the NPCs you are interested in. I am sure we can make that relatively convenient, except for the time that it takes to start up OpenMW which should not be a problem on any machine that wasn't made from a recycled toaster.

**Comment by scrawl** [comment]
Far ahead. Very.

To get any sort of useful autocalc in a script, we'd need:
* Scripting functions that don't exist yet (store access, record field access, etc.)
* New scripting data structures and constructs (associative containers, for loops, etc)

There is still the question of extending the original scripting language or using an existing language. 

Then we need to start actually writing those scripts, which is kinda a rewrite of the OpenMW codebase itself. And we shouldn't rush to this stage. If we got one of the prerequisites wrong, we'll have to rewrite all the scripts <i>again</i>. Then we need to regression test to make sure nothing broke along the way.

I don't mean to be a downer, but I think you are seriously underestimating the work involved. Even if we start immediately after 1.0 is released, it would probably land in time for 1.3 or 1.4. Also, I think you are underestimating how long it will take to get to 1.0. I guess we are 90% of the way there feature wise, but there are still lots of bugs to squash, which usually takes longer than adding those features in the first place.

> Fire up OpenMW with a debug profile that sets up the situation you want to review. Inspect the NPCs you are interested in. 

That sounds seriously annoying. You'd need a new debug profile for every NPC you want to inspect. You'd need to use the console to inspect values, which is less convenient than a GUI. Starting the game does take a few seconds, even on a powerful rig.

**Comment by zinnschlag** [comment]
> I don't mean to be a downer, but I think you are seriously underestimating the work involved. Even if we start immediately after 1.0 is released, it would probably land in time for 1.3 or 1.4.

I would be able to get all required script enhancements done within a week if I focused exclusively on OpenMW during that week. Yeah, I know, not going to happen. But still the situation isn't that bad.
And the code for the scripts already exists and only needs to be translated from C++.
Of course we need a lot of regression testing. I agree that we should be careful not to mess it up, but overall I think you are to pessimistic.

For being feature complete 1.4 sounds about right. But even that isn't so far into the future that we shouldn't consider it. With not having a fixed blueprint anymore after 1.0 (a.k.a vanilla Morrowind) I think it is fairly important to plan and think ahead a bit farther. 

> That sounds seriously annoying. You'd need a new debug profile for every NPC you want to inspect. You'd need to use the console to inspect values, which is less convenient than a GUI. Starting the game does take a few seconds, even on a powerful rig.

Of course you would not need a debug profile for each NPC. Just one for each factor and situation that affects the outcome of auto calc.
As for using the console, the plan is to utilise OpenMW-CS as a debugging tool. Eventually you should be able to run the entire thing from the OpenMW-CS GUI with OpenMW just working in the background.
